Add tests for various System.Drawing.Common brushes

hughbe opened the pull request:
And cleanup some code

hughbe commented:
Test Innerloop OSX10.12 Release x64 Build and Test

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Brush.cs:
```diff
@@ -100,34 +60,13 @@ protected virtual void Dispose(bool disposing)
-                    }
-
-                    Debug.Fail("Exception thrown during Dispose: " + ex.ToString());
-                }
```
Why is it safe to remove this catch block?

hughbe replied:
The code for checking the exception is
```cs
public static bool IsCriticalException(Exception ex)
{
    return ex is NullReferenceException
            || ex is StackOverflowException
            || ex is OutOfMemoryException
            || ex is System.Threading.ThreadAbortException
            || ex is ExecutionEngineException
            || ex is IndexOutOfRangeException
            || ex is AccessViolationException;
}

public static bool IsSecurityOrCriticalException(Exception ex)
{
    return (ex is System.Security.SecurityException) || IsCriticalException(ex);
}
```

SecurityExceptions are not going to be thrown in .NET Core, as the coreclr doesn't support these security features.

The only other option that can be thrown from a delete is an AccessViolationException, which can occur in the following code example:
```
[Fact]
public void SetNativeBrush_Brush_Success()
{
    using (var brush = new SubBrush())
    {
        brush.PublicSetNativeBrush((IntPtr)10);
    }
}

private class SubBrush : Brush
{
    public override object Clone() => this;
    public void PublicSetNativeBrush(IntPtr brush) => SetNativeBrush(brush);
}
```

It's my understanding that GDI+ can't throw CLR exceptions, so only an AV can occur, which will already be handled.

stephentoub replied:
What if it gets another runtime-related exception type, like EntryPointNotFound for the P/Invoke?  Or what about when Mono wants to use this implementation rather than the current one?

hughbe replied:
thats a fair point about entrypointnotfound. We could get this in Windows Nano

I thought that the debug.fail indicates that we really really don't expect any exception, but we've seen some fault assertions in the library already as it seems no one ever tested it in debug mode.

Do you reckon I should revert this? I didn't quite understand your comment about mono.

stephentoub replied:
> Do you reckon I should revert this?

Yes.

> I didn't quite understand your comment about mono.

The Mono runtime may behave differently in various situations from CoreCLR, so your assertions about behaviors of CoreCLR and what exceptions the runtime may or may not throw (e.g. AccessViolationException) may or may not hold if this code is used on Mono.  And the plan is for more and more CoreFx code to be used to replace code in Mono's libraries.

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Drawing2D/HatchBrush.cs:
```diff
@@ -2,137 +2,69 @@
-            if (status != SafeNativeMethods.Gdip.Ok)
-                throw SafeNativeMethods.Gdip.StatusException(status);
+        { 
+            int status = SafeNativeMethods.Gdip.GdipCreateHatchBrush(unchecked((int)hatchstyle), foreColor.ToArgb(), backColor.ToArgb(), out IntPtr brush);
```
In cases like this, I'd actually prefer it if we didn't use the new C# syntax and left the variable defined on its own line.  Otherwise, it's too difficult when reading the code to see where brush is actually declared.

hughbe replied:
Fair enough - I reverted this change and others

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/Drawing2D/HatchBrush.cs:
```diff
@@ -2,137 +2,69 @@
-            IntPtr cloneBrush = IntPtr.Zero;
-
-            int status = SafeNativeMethods.Gdip.GdipCloneBrush(new HandleRef(this, NativeBrush), out cloneBrush);
+            int status = SafeNativeMethods.Gdip.GdipCloneBrush(new HandleRef(this, NativeBrush), out IntPtr cloneBrush);
```
Same out IntPtr comment.  Will stop commenting on these.

hughbe replied:
Reverted

stephentoub commented on src/System.Drawing.Common/src/System/Drawing/SolidBrush.cs:
```diff
@@ -2,44 +2,26 @@
-    /// </devdoc>
-    public sealed class SolidBrush : Brush, ISystemColorTracker
+    public sealed class SolidBrush : Brush
+#if FEATURE_SYSTEM_EVENTS
```
Why are you adding this define?  If we don't need the code, we should remove it entirely.  If we do need it, we shouldn't have it conditionally compiled.

hughbe replied:
I went ahead and removed the unused FEATURE_SYSTEM_EVENTS flag

stephentoub commented on src/System.Drawing.Common/tests/BrushTests.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace System.Drawing.Tests
+{
+    public class BrushTests : RemoteExecutorTestBase
```
Why RemoteExecutorTestBase?

hughbe replied:
Vestigal - removed

danmosemsft commented on src/System.Drawing.Common/tests/Drawing2D/HatchBrushTests.cs:
```diff
@@ -0,0 +1,126 @@
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        public void ForegroundColor_EmptyAndGetDisposed_ThrowsArgumentException()
+        {
+            var brush = new HatchBrush(HatchStyle.DarkHorizontal, Color.PeachPuff, Color.Purple);
```
`using (var brush = new HatchBrush(HatchStyle.DarkHorizontal, Color.PeachPuff, Color.Purple)) { }`
may be preferable style?

hughbe commented:
Test Portable Linux x64 Debug Build
Test Portable Linux x64 Release Build

(no logs)

hughbe commented:
Rebased to fix merge conflicts and the CI glitches from the recent changes to test names

hughbe commented:
Test Linux x64 Release Build
Test OSX x64 Debug Build

hughbe commented:
Test Linux x64 Release Build

stephentoub commented on src/Common/src/System/Drawing/KnownColorTable.cs:
```diff
@@ -56,9 +52,6 @@ private static void InitColorTable()
 
-#if FEATURE_SYSTEM_EVENTS
-            SystemEvents.UserPreferenceChanging += new UserPreferenceChangingEventHandler(OnUserPreferenceChanging);
-#endif
```
@mellinoe, are you ok with removing all of this FEATURE_SYSTEM_EVENTS stuff?  Is there a plan to enable it?  Will removing it mess up Mono's ability to source share?

mellinoe replied:
I would prefer to keep the code here until we decide we can't or don't want to ever reactivate it. Let's discuss it over at #21319

mellinoe commented on src/System.Drawing.Common/src/System/Drawing/Gdiplus.cs:
```diff
@@ -2627,6 +2627,12 @@ internal static int GdipDeleteStringFormat(HandleRef format)
+            internal static void CheckStatus(int status)
+            {
+                if (status != Ok)
+                    throw StatusException(status);
```
nit: Please add braces around this

mellinoe commented on src/System.Drawing.Common/src/System/Drawing/SolidBrush.cs:
```diff
@@ -99,28 +59,19 @@ protected override void Dispose(bool disposing)
-
-                    if (status != SafeNativeMethods.Gdip.Ok)
-                        throw SafeNativeMethods.Gdip.StatusException(status);
+                    int status = SafeNativeMethods.Gdip.GdipGetSolidFillColor(new HandleRef(this, NativeBrush), out int colorARGB);
```
nit: move the variable declaration out do be consistent.

hughbe commented:
I've addressed your PR feedback. I reverted the commit removing FEATURE_SYSTEM_EVENTS. I expected it to have some discussion anyway. You won't see the actual revert in the PR feedback because I just went back and removed the commit

hughbe commented:
Test Linux x64 Release Build
Test Windows x64 Debug Build
